# Data from the 1000 genomes project

Data presented here are information about allele frequencies for 3 different populations for all of chromosome 6.

The dataset is ~15000000 rows (~5000000 SNPs pr population)

The columns are

"position"				position on chromosome 6
"reference_allele"      the DNA base (ACG or T) of the reference allele 
"alternative_allele"    the DNA base (ACG or T) of the alternative allele 
"population"            the population, AFR, EUR or EAS (Africa, Europe og East Asia)
"frequency"             the alternative allele frequency

# Stop reading - the rest is purely technical stuff if I need to do it again.

# Getting the data from the web

# Download VCF for chromosome 6
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr6_GRCh38_sites.20170504.vcf.gz

# Only keep biallelic SNPs
source /com/extra/bcftools/1.3/load.sh
bcftools view  -O z --max-alleles 2 --types snps ALL.chr6_GRCh38_sites.20170504.vcf.gz >tmp.vcf.gz

# Get position, ref and alt allele
zcat tmp.vcf.gz | grep -v "#" | cut -f 2,4,5 >tmp1

# Extract information from different populations
zcat tmp.vcf.gz | grep -v "#" | cut -f 8 | cut -f 6 -d ";" | cut -f 2 -d "=" | awk -v "OFS=\t" '{print "EAS",$0}' >tmp2
zcat tmp.vcf.gz | grep -v "#" | cut -f 8 | cut -f 7 -d ";" | cut -f 2 -d "=" | awk -v "OFS=\t" '{print "AMR",$0}' >tmp3
zcat tmp.vcf.gz | grep -v "#" | cut -f 8 | cut -f 8 -d ";" | cut -f 2 -d "=" | awk -v "OFS=\t" '{print "AFR",$0}' >tmp4
zcat tmp.vcf.gz | grep -v "#" | cut -f 8 | cut -f 9 -d ";" | cut -f 2 -d "=" | awk -v "OFS=\t" '{print "EUR",$0}' >tmp5
zcat tmp.vcf.gz | grep -v "#" | cut -f 8 | cut -f 10 -d ";" | cut -f 2 -d "=" | awk -v "OFS=\t" '{print "SAS",$0}' >tmp6

# paste together
paste tmp1 tmp2 >tmp.pop.EAS
paste tmp1 tmp3 >tmp.pop.AMR
paste tmp1 tmp4 >tmp.pop.AFR
paste tmp1 tmp5 >tmp.pop.EUR
paste tmp1 tmp6 >tmp.pop.SAS

# Put it after eachother - to make tidy data - one row is one observation
# We only use 3 populations to save space
# EURope, AFRica, EASt asia

cat tmp.pop.EUR tmp.pop.AFR tmp.pop.EAS | gzip -c9 >tmp.tsv.gz

# Resulting headers are
# POS, REF, ALT, POP, FREQ

# In R

library(tidyverse)


d = read_delim(file="tmp.tsv.gz",
               delim=c("\t"), 
               col_names = c("position", "reference_allele", "alternative_allele", "population", "frequency"))

# Identify positions where it is fixed in all populations

# Fixed for reference allele
bad1 = d %>% filter(frequency==0 & population=="EUR") %>% select(position) %>%
  intersect(d %>% filter(frequency==0 & population=="AFR") %>% select(position)) %>%
  intersect(d %>% filter(frequency==0 & population=="EAS") %>% select(position))

# Fixed for alternative allele
bad2 = d %>% filter(frequency==1 & population=="EUR") %>% select(position) %>%
          intersect(d %>% filter(frequency==1 & population=="AFR") %>% select(position)) %>%
          intersect(d %>% filter(frequency==1 & population=="EAS") %>% select(position))

badsnps = rbind(bad1,bad2)

# Filter to only keep SNPs not in the bad1 or bad2 set
d2 = d %>% filter(!position %in% badsnps$position)

write_delim(d2, path="1000g.allele.frequencies.tsv.gz", delim = "\t")

